### ISSUE TITLE
Flex Attention Decoding Silently Incorrect with H=None Broadcasting and Sliced mask_mod

### ISSUE THREAD

**Issue by zaptrem** [pytorch/pytorch#163314]
## 🐛 Describe the bug

Two bugs in `torch.nn.attention.flex_attention` cause silent incorrect results when used for autoregressive decoding with BlockMask slicing.

### Bug 1: H=None Broadcasting Produces Incorrect Results
When creating a BlockMask with `H=None` (which should broadcast to any head count), the attention computation produces silently incorrect results.

```python
# This should work but produces wrong results
block_mask = create_block_mask(causal_mask, B=4, H=None, Q_LEN=512, KV_LEN=512, ...)
```

**Expected:** H=None should broadcast correctly to match the actual number of attention heads
**Actual:** Results differ significantly from ground truth (max error ~2.77)

### Bug 2: Sliced BlockMask mask_mod Becomes Noop
After slicing a BlockMask, the `mask_mod` attribute is replaced with a noop function. If users attempt to modify the sliced mask's `mask_mod` directly (instead of using the base mask's `mask_mod`), results are silently incorrect.

```python
base_mask = create_block_mask(...)
sliced_mask = base_mask[:, :, block_offset]
# sliced_mask.mask_mod is now noop_mask, not the original mask function!
sliced_mask.mask_mod = get_mask_mod_with_offset(sliced_mask.mask_mod, offset)  # WRONG - uses noop
sliced_mask.mask_mod = get_mask_mod_with_offset(base_mask.mask_mod, offset)    # CORRECT
```

**Expected:** Clear documentation or error when mask_mod is incorrectly modified (or just maintain the correct mask mod function)
**Actual:** Silent incorrect results (max error ~0.18)

## Impact
These bugs affect autoregressive generation use cases where BlockMask slicing is used for efficient decoding, as described in the [PyTorch blog post on flex decoding](hhttps://pytorch.org/blog/flexattention-for-inference/).

## Reproduction

https://gist.github.com/zaptrem/bd47b71fa2908ebbc5b2a26c6fcf5070

Test script attached demonstrates all three scenarios:
1. ✅ Correct implementation (H=8, base mask_mod)
2. ❌ H=None broadcast bug (max error: 2.77)
3. ❌ Wrong mask_mod application bug (max error: 0.18)

All scenarios are validated against `F.scaled_dot_product_attention` as ground truth.

## Versions

```
PyTorch version: 2.8.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.11.13 (main, Jun  4 2025, 08:57:29) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.5.0-1023-aws-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-80GB
Nvidia driver version: 535.183.01
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @bdhirsh @penguinwu @Chillee @drisspg @yanboliang @BoyuanFeng @chauhang @ydwu4

**Comment by drisspg** [comment]
Second issue is intentional

But the first one looks like a real bug

**Comment by drisspg** [comment]
We don't have any 'state' that is easily checkable to know when a BM is post slicing or just any other BM so I am not quite sure where that check would go

**Comment by zaptrem** [comment]
> We don't have any 'state' that is easily checkable to know when a BM is post slicing or just any other BM so I am not quite sure where that check would go

Could simply replace the noop_mask_mod with a mask_mod that simply throws an error that tells the user the right way to use slicing.
